jerry-core/ecma/operations/ecma-regexp-object.c

```diff
@@ -508,31 +464,159 @@ ecma_regexp_canonicalize_char (lit_code_point_t ch) /**< character */
+    {
+      return (cp < LIT_CHAR_0 || cp > LIT_CHAR_9);
+    }
+    case RE_ESCAPE_WORD_CHAR:
+    {
+      return lit_char_is_word_char (cp);
+    }
+    case RE_ESCAPE_NOT_WORD_CHAR:
+    {
+      return !lit_char_is_word_char (cp);
+    }
+    case RE_ESCAPE_WHITESPACE:
+    {
+      return lit_char_is_white_space ((ecma_char_t) cp);
+    }
+    case RE_ESCAPE_NOT_WHITESPACE:
+    {
+      return !lit_char_is_white_space ((ecma_char_t) cp);
+    }
+    default:
+    {
+      JERRY_UNREACHABLE ();
+    }
+  }
+} /* ecma_regexp_check_class_escape */
+
+/**
+ * Helper function to get current code point or code unit depending on execution mode,
+ * and advance the string pointer.
+ *
+ * @return lit_code_point_t current code point
+ */
+static lit_code_point_t
+ecma_regexp_advance (ecma_regexp_ctx_t *re_ctx_p, /**< regexp context */
+                     const lit_utf8_byte_t **str_p) /**< reference to string pointer */
+{
+  JERRY_ASSERT (str_p != NULL);
+  lit_code_point_t cp = lit_cesu8_read_next (str_p);
+
+#if ENABLED (JERRY_ES2015)
+  if (JERRY_UNLIKELY (re_ctx_p->flags & RE_FLAG_UNICODE)
+      && lit_is_code_point_utf16_high_surrogate ((ecma_char_t) cp)
+      && *str_p < re_ctx_p->input_end_p)
+  {
+    const ecma_char_t next_ch = lit_cesu8_peek_next (*str_p);
+    if (lit_is_code_point_utf16_low_surrogate (next_ch))
+    {
+      cp = lit_convert_surrogate_pair_to_code_point ((ecma_char_t) cp, next_ch);
+      *str_p += LIT_UTF8_MAX_BYTES_IN_CODE_UNIT;
+    }
+  }
+#endif /* ENABLED (JERRY_ES2015) */
+
+  return ecma_regexp_canonicalize (cp, re_ctx_p->flags);
+} /* ecma_regexp_advance */
+
+#if ENABLED (JERRY_ES2015)
+/**
+ * Helper function to get current full unicode code point and advance the string pointer.
+ *
+ * @return lit_code_point_t current code point
+ */
+lit_code_point_t
+ecma_regexp_unicode_advance (const lit_utf8_byte_t **str_p, /**< reference to string pointer */
+                             const lit_utf8_byte_t *end_p) /**< string end pointer */
+{
+  JERRY_ASSERT (str_p != NULL);
+  const lit_utf8_byte_t *current_p = *str_p;
+
+  lit_code_point_t ch = lit_cesu8_read_next (&current_p);
+  if (lit_is_code_point_utf16_high_surrogate ((ecma_char_t) ch)
+      && current_p < end_p)
+  {
+    const ecma_char_t next_ch = lit_cesu8_peek_next (current_p);
+    if (lit_is_code_point_utf16_low_surrogate (next_ch))
+    {
+      ch = lit_convert_surrogate_pair_to_code_point ((ecma_char_t) ch, next_ch);
+      current_p += LIT_UTF8_MAX_BYTES_IN_CODE_UNIT;
+    }
+  }
+
+  *str_p = current_p;
+  return ch;
+} /* ecma_regexp_unicode_advance */
+#endif /* ENABLED (JERRY_ES2015) */
+
+/**
+ * Helper function to revert the string pointer to the previous code point.
+ *
+ * @return pointer to previous code point
+ */
+static JERRY_ATTR_NOINLINE const lit_utf8_byte_t *
+ecma_regexp_step_back (ecma_regexp_ctx_t *re_ctx_p,
+                       const lit_utf8_byte_t *str_p) /**< reference to string pointer */
```

Comment:
When this function is exactly used? Word boundary?

If there is a surrogate pair, and match starts from the position between the two characters, `ecma_regexp_step_back` may go back the start of the surrogate pair instead of stopping at the middle.

Reply:
It is used in the greedy atom iterator during backtracking.

In unicode mode surrogate pairs are required to be handled as a single code point, matching can never start from the middle of a surrogate pair, and it is required to read the entire surrogate pair when backtracking.

Reply:
It seems you are right, The engine overwrites the start offset for surrogate pairs. Is there a test for this?

Reply:
There is one now.